### DISCUSSION TITLE
Set CompatibilityMode to STRICT

### DISCUSSION THREAD

**Discussion by m4rkyx** [fluentmigrator/fluentmigrator#1485] [Q&A]
Hello guys,

As we saw [here](https://github.com/fluentmigrator/fluentmigrator/issues/812#issuecomment-383417163), there is a comptability mode that we can set to STRICT (default is LOOSE) when applying migration.

We're using SQLite and there is some limitations that we need to handle in our migrations.
Currently, FluentMigrator does not apply all of our migrations but doesn't say anything.

So is it possible to stop the migration process when a SQL statement is in error or unhandled by the database provider ?

Thanks a lot :)

PS : The last version of SQLite 3.35.0 support ALTER TABLE DROP COLUMN, are you planning to implement it quickly ?

**Comment by jzabroski** [comment]
@m4rkyx Hi, thanks for discussing this.

Is your proposed pseudo-code:

1. Call `select sqlite_version();` and save it to variable `sqliteVersion`.
2. If `sqliteVersion >= 3.35.0` then allow drop column else (what do you expect to have happen?)

Separately, it seems CompatiblityMode should be set to STRICT.  I don't understand from reading the discussion you linked why the default is LOOSE.  That behavior seems chaotic to me.  It would be a breaking change, but a sensible one that could avoid disastrous results.  @fubar-coder Any idea why we would keep the default to be LOOSE?  Am I not thinking of something?

**Reply by fubar-coder** [reply]
I guess that (before I took over) the previous maintainer took a pragmatic approach: Everything should just work for the most common operations. Maybe it'd be too confusing to get SQLite support working with STRICT as default behavior, because it'd require the FM users to read documentation.

**Reply by jzabroski** [reply]
Hmm... I think that it probably requires a tiny bit more research before we go fix.

@m4rkyx I think we need to figure out what the LOOSE behavior silently allows for now before we fix this.  Would you volunteer organizing some documentation and maybe some tests?

I regret that the SQLite tests are Facts and not Theories parameterized by CompatibilityMode.

**Reply by m4rkyx** [reply]
Hi,
Thanks for your follow-up :)

I think, basically, we want to know when an expression is not supported by the provider.
All of the expressions [here](https://github.com/fluentmigrator/fluentmigrator/blob/9c87d698b161b3a1bb596a5624c02891a64459c5/src/FluentMigrator.Runner.SQLite/Generators/SQLite/SQLiteGenerator.cs) that are not supported should throw an exception.

Instead, code below this will run without exception by making us believe that everything was done correctly when it was not. (colom was not dropped)
![image](https://user-images.githubusercontent.com/1868168/122520553-0662e580-d014-11eb-90a1-a37f22f40880.png)

Maybe I'll have some times in a few days to create documentation and tests if it's still needed.

Thanks again !

**Reply by jzabroski** [reply]
A Lesson Learned for FluentMigrator is that we handled newer versions of the raw SQL text generation via specifying a processorId.  ProcessorId and the Processor interface are not tightly coupled to the Generator, per say, but de facto they are.  And the ProcessorId for the most part is tightly coupled to a particular database driver library.  And, historically, database systems would release a new database driver library in conjunction with new SQL language features. - But it doesn't have to be that way.  For example, SQL Server's minimum requirement is that the SQL command you issue be speak-able in its Tabular Data Stream (TDS) Protocol.  Several major versions of their database driver libraries speak the same Protocol version.

Netting it out, I wish we had a nicer way to modularly customize the generation logic based on language version/protocol version, and had some table that had a self-describing list of which features were available per version.  But that gets fairly hairy fairly quickly.  As an example, the Rename operation in SQL Server is supported, but may not work depending on bugs introduced to their rename tool under special conditions (one version of the product shipped with broken rename support for temporal tables iff a base column was dropped and another added).

**Reply by jzabroski** [reply]
@m4rkyx After thinking this over, I think the SQLiteGenerator is pretty much trash :)  I don't know why it has so many HandleCompatibility calls for things that have been in SQLite even since 2001.  I'm sorry I was obtuse - I didn't realize so many features SQLite has supported for years.

**Reply by jzabroski** [reply]
@m4rkyx To Clarify, I think the right approach is to rewrite the SQLiteGenerator and deprecate the old one.  People who want the old one can still use it.

Open to suggestions on what to call the new class.  Probably the right thing to do is just drop the old class and move to FluentMigrator V4 code base ASAP.
